Specialize Optional[T] to T (or subtype for Tensor) or None when executing graph

t-vi opened the pull request:
This patch specializes `Optional[Tensor]` graph inputs to either a `DimensionedTensorType` (if a Tensor is passed) or `NoneType`. Other `Optional[T]` are specialized to `T` or `None`.

## Functional changes
- For unwrapping (checked and unchecked) we need to keep the output type, as IR code that follows unwrapping may not work with NoneType (just as it doesn't deal with Optional). While it would not be hit during execution, it will run against the (legitimate) assumptions of the analysis passes.
- Function lookup currently will not match NoneType when it expects optional (I'm not entirely sure why this doesn't lead to unhappyness currently, but hey), I amend this at the level of the function matching code (`operator.cpp`), but see Adam's comments. We would run into trouble if we needed to select between functions whose signature only differs in Optional types with different subtypes, but we would have the same problem when calling them directly, so I would think this is OK.

## Advantages:
- It would enable throwing away branches we can't hit. This also reduces the "blockyness" of the graph, so it may be easier to apply optimizations (e.g. fuse things in `if t is None: ...` and outside the `if`.
- Arguments passed into `Optional[Tensor]` arguments will get shape information, which is very handy.
- It get's rid of the problem that tensors passed into Optional arguments get requires_grad set erroneously #18270 (though that also affects lists, which aren't fixed here).
- `Optional[List[int]]` is needed for #18697.

## Disadvantages:
- We're changing typing in a more subtle way than the `TensorType`->`DimensionedTensorType`.
- In particular, specializing to NoneType loses the Type information captured in the `OptionalType` element type.

t-vi committed:
Specialize Optional (Tensor) to None when executing graph

In #18360, we used undefined Tensor (aka AutogradZeroTensor),
but this can be errorprone when the type or value is compared
to None, e.g. as seen when comined with the (not yet landed)

For this to work, we must allow None passed to functions
taking Tensor?.

t-vi committed:
don't unpack NoneType in _unwrap_optional

apaszke commented on torch/csrc/jit/fuser/fallback.cpp:
```diff
@@ -35,7 +35,8 @@ void runFallback(int64_t key, Stack& stack) {
     throw std::runtime_error("Failed to find fusion spec to run fallback.");
-
+  // std::cout << "fallback!" << std::endl;
+  //(*maybe_spec)->graph()->dump();
```
Please remove!

t-vi replied:
Oh oh. :+1:

apaszke commented on torch/csrc/jit/passes/shape_analysis.cpp:
```diff
@@ -1014,11 +1013,9 @@ class ShapePropagator {
-    //   device         : preserved
-    //   tensor inputs  : 1
-    //   tensor outputs : 1
+    //   dims           : 0 if dim is None, otherwise preserved if keepdim ==
```
This comment is messed up now

t-vi replied:
Sorry!

apaszke commented on torch/csrc/jit/operator.cpp:
```diff
@@ -457,10 +457,20 @@ bool Operator::matches(const Node* node) const {
+          formals[i].type()->kind() == TypeKind::OptionalType) {
+        // when looking for a match, None is actually OK here
+        continue;
+      }
```
Uh how come do we need this special case here, and how did we survive without it before? We should be able to unify a `NoneType` with an `OptionalType`, right?

t-vi replied:
Well, NoneType isn't in a Subtype relation with OptionalType, so we won't find signatures. Before we had an OptionalType going in.

apaszke replied:
Hmm ok, I guess we could change it? Those conditionals are worrying for me. I'm not really sure how our type hierarchy looks right now, and if it wouldn't mess up some things. cc @zdevito @eellison @wanchaol

t-vi replied:
I added it because looked like it's intentional: https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/core/jit_type.h#L221

eellison replied:
I think this is fine as well but maybe @zdevito should weigh in

apaszke replied:
Our `None` story is extremely confusing to be honest. The comment you linked says this as well
> instead NoneType is convertable in schema matching to any Optional[T]

which seems to imply that what you're doing in here should have been already present in the code?

I'd really like to hear what @zdevito thinks about this because I haven't been fully following the `None` changes, but it looks very weird to me.

zdevito replied:
This code cannot be modified in this way. NoneType values should not appear in the graph _anywhere_ after the compiler has run. All None's need to be specialized to a particular type otherwise logic related to unwrapping options will no longer work.  Allow None here will cause type variable matching to fail.

apaszke commented on torch/csrc/jit/passes/shape_analysis.cpp:
```diff
@@ -514,14 +514,13 @@ class ShapePropagator {
+        } else if (!node->input()->type()->isSubtypeOf(NoneType::get())) {
+          node->output()->setType(node->input()->type());
+        }
+        return;
```
Is this just formatting? Can you please avoid doing formatting only changes?

t-vi replied:
No it's not. I check for NoneType now to avoid propagating a None outside.
I try to. Here I actually messed up the formatting before and just re-formatting the changed line would have made things worse.

apaszke commented on torch/csrc/jit/passes/shape_analysis.cpp:
```diff
@@ -539,17 +538,17 @@ class ShapePropagator {
-	return;
+        if (auto ot = node->input()->type()->cast<OptionalType>()) {
+          node->output()->setType(ot->getElementType());
+        } else if (!node->input()->type()->isSubtypeOf(NoneType::get())) {
```
Why?

t-vi replied:
During analysis, we may have None going in when we specialize the graph. This will be in a branch that is later pruned / not executed, but it'll screw the analysis.
I'll add this as a code comment.

apaszke replied:
Why would it screw the analysis? If we know it will deleted later, why does it matter?

t-vi replied:
Well, the code following the unwrap will need a containing type. So we'll hit asserts if we continue with None.

t-vi replied:
So the None part is gone now, and below, Zach has comments on the same place.

apaszke commented on torch/csrc/jit/passes/specialize_autogradzero.cpp:
```diff
@@ -16,7 +16,8 @@ void specializeAutogradZero(Graph& g) {
     const auto& tp = input->type();
-    if (tp->isSubtypeOf(AutogradZeroTensorType::get())) {
+    if (tp->isSubtypeOf(AutogradZeroTensorType::get()) ||
+        tp->isSubtypeOf(NoneType::get())) {
```
This is unnecessary. We never use `Optional[Tensor]` for those inputs, so it will never get triggered. It might actually be a good idea to get rid of `AutogradZeroTensorType` though (in favor of optionals).

t-vi replied:
Well. I get failures without it. :confused:

apaszke replied:
What kinds of failures? I would really like us to understand what's happening instead of applying patches blindly

t-vi replied:
So what happens is for Optional tensor arguments and None inputs, the `ArgumentInfo` (`argument_spec.h`) will have `defined=false`. If now the backward graph takes an `OptionalType` argument, we specialize to `None` (instead of `AutogradZeroTensorType`).
But this then confuses the above if-statement, which differentiates between not-passed-in Tensors and passed-in-Tensors, until now by using `AutogradZeroTensorType`.
In a way, we're involuntarily taking a tiny step towards getting rid of `AutogradZeroTensorType` (which would involve making the replacement in the `ArgumentSpec` unconditional and then one could probably remove the test for `AutogradZeroTensorType` above).

t-vi replied:
This is gone now with better None vs. undefined handling.

t-vi committed:
fix worst stuff

t-vi commented:
Given the rather fundamental nature of the shortcomings of this approach, I'd have a tendency to drop it.

t-vi committed:
Merge branch 'master' into optional_None

t-vi committed:
Merge branch 'master' into optional_None

eellison commented:
Is this the relevant PR for specializing Optional tensors as graph input?

t-vi commented:
Yes.

t-vi committed:
Fixes

eellison commented on test/test_jit.py:
```diff
@@ -4572,6 +4572,39 @@ def named_var_and(x, y):
+        res = fn(t)
+        self.assertEqual(res, 0)
+        g = fn.graph_for(t)
+        self.assertEqual(list(g.inputs())[0].type().kind(), 'DimensionedTensorType')
```
fwiw you can write `next` instead of `list(...)[0]`

eellison commented on test/test_jit.py:
```diff
@@ -4572,6 +4572,39 @@ def named_var_and(x, y):
+
+        res = fn(t)
+        self.assertEqual(res, t)
+        g = fn.graph_for(t)
```
doesn't look like this g is being used

eellison commented on torch/csrc/jit/argument_spec.h:
```diff
@@ -153,10 +153,16 @@ struct ArgumentSpec {
-      if (!arg.defined())
-        return AutogradZeroTensorType::get();
+      if (!arg.defined()) {
+        if (original->isSubtypeOf(OptionalType::ofTensor())) {
```
Could you add a comment here?

apaszke replied:
Ok, this actually looks wrong to me. Basically when we return an `Optional[Tensor]` from forward, then there are 3 cases to consider for gradients:
1. The result was `None`. It's not a tensor, so it cannot be differentiated.
2. The result was unused/gets no gradient, so it should be treated as a zero tensor.
3. The result was used/gets a gradient, so it is non-zero and we shouldn't optimize it out.

This `if` does not seem to distinguish 1. and 2., which later results in some very awkward code you had to put into AD code, and I would very much like to see those parts gone. A good way to start fixing it is to start differentiating between them in here.

t-vi replied:
This is now fixed to distinguish between all three. :)

eellison commented on torch/csrc/jit/ir.cpp:
```diff
@@ -777,8 +777,8 @@ bool Node::matches(
       (kind_ == prim::Constant && !this->hasAttributes() &&
-       (output()->type()->cast<OptionalType>() ||
-        output()->type() == NoneType::get()));
+       output()->type()->cast<OptionalType>()) ||
```
Good fix!

t-vi replied:
:slightly_smiling_face:

t-vi replied:
Actually, very bad breakage! It throws if we have several outputs. That case needs to be covered in `Value::mustBeNone`.

eellison commented on torch/csrc/jit/passes/peephole.cpp:
```diff
@@ -188,16 +188,24 @@ void PeepholeOptimizeImpl(Block* block, bool addmm_fusion_enabled) {
-          auto output = node->owningGraph()->insertConstant(
-              node->kind() == aten::__isnot__);
-          node->output()->replaceAllUsesWith(output);
+      if (node->inputs().at(0)->mustBeNone() &&
```
This works and we probably don't want to further complicate the PR... I think the more general case is in change constant propagation so that it knows `prim::AutogradZero` is a constant, but i'm not sure how much that would buy us in optimization anyways. you can ignore this comment lol

t-vi replied:
Well, I'm not sure an argument is necessarily a constant (though arguably None is), so constant propagation feels a bit wrong here. But I'm very happy to move it.

eellison replied:
I think it's good in

eellison commented on torch/csrc/jit/passes/shape_analysis.cpp:
```diff
@@ -529,10 +540,20 @@ class ShapePropagator {
       case aten::_unwrap_optional: {
+        // if we have None as input, we need to leave the output alone
         auto input_ivalue = toIValue(node->input());
         if (input_ivalue && input_ivalue->isNone()) {
```
Nit: Could you use the mustBeNone api here instead ?

t-vi replied:
Are you certain that that works? I'm happy to do it, but I do rely on your judgement here, that it covers all the cases we need. (And the tests might show.)

eellison replied:
Pretty sure. The only additional case that it is covering is prim::AutoGradZero, in which case we're not gaining anything by continuing with the other two cases right?

apaszke replied:
Yeah it looks like a candidate for `mustBeNone()`

t-vi replied:
Done now with the amended mustBeNone.

eellison commented on torch/csrc/jit/passes/shape_analysis.cpp:
```diff
@@ -529,10 +540,20 @@ class ShapePropagator {
+	// with the nodes consuming the output.
+        if (auto ot = node->input()->type()->cast<OptionalType>()) {
+          node->output()->setType(ot->getElementType());
+        } else if (!node->input()->type()->isSubtypeOf(NoneType::get())) {
```
Nit: Maybe rewrite `!node->input()->type()->cast<NoneType>` so that it follows a little more closely with your previous check

t-vi replied:
I think if we use mustBeNone, we have both covered, so this can be a plain else.

eellison reviewed (commented):
Looks good to me, just a few minor nits. I think someone with a better knowledge of the specialize autogradzero pass should review before we land @apaszke @zdevito

apaszke commented on torch/csrc/jit/passes/shape_analysis.cpp:
```diff
@@ -513,6 +513,17 @@ class ShapePropagator {
+        } else if (!node->input()->type()->isSubtypeOf(NoneType::get())) {
+          node->output()->setType(node->input()->type());
+        }
+        return;
```
When do we emit `unchecked_unwrap_optional`s though? It seems that if they're unchecked then they should never see `NoneType`s as inputs, or they would silently result in UB at runtime? I don't understand what does it mean to "keep the graph consistent". This simply looks wrong

t-vi replied:
Well, we know that the if branch containing `unchecked_unwrap_optional` will never be executed. However, the branch is still there during the various JIT passes.

t-vi replied:
They don't see None anymore.

eellison commented:
I'm not sure that this case will necessarily be hit but just to be safe could you modify the operator of `aten::_unwrap_optionnal` so that it checks for the undefined tensor case ?

apaszke reviewed (dismissed):
I think I finally understand where a lot of confusion is coming from, and why AD has regressed in weird ways. See my comment in `argument_spec.h`.

t-vi committed:
incorporate some feedback from eellison, thanks

eellison commented on torch/csrc/jit/ir.cpp:
```diff
@@ -639,7 +639,7 @@ std::shared_ptr<Graph> Graph::copy() {
 
 bool Value::mustBeNone() const {
-  return node_->mustBeNone();
+  return type() == NoneType::get() || node_->mustBeNone();
```
Why was this needed ?

t-vi replied:
Because of the "bad breakage" above, we can't check for None parameters in `node_->mustBeNone`.
For parameters, you have (in general) more than one output.

t-vi replied:
Gone now. :)

t-vi committed:
Merge branch 'master' into optional_None

t-vi commented:
So after merging master (including the `ArgumentSpecCreator`), I do distinguish between undefined tensor and None by having an extra flag `is_none_` instead of abusing `defined_`. As the `is_tensor_` is dropped, I added a new `is_nontensor_optional_` (so that zeroing the struct will cover the tensor case).
I actually have the non-tensor case (that used to be part of #18697) here, maybe they're better here.

Note that before this patch, `Optional[Tensor]` hadn't been specialized at all, now we specialize into cases for the various dimensions and `None`. On the other hand `Optional[Tuple[...]]` and `Optional[Class[...]]` only looked at in terms of None / Value, not in terms of their contents (but that wasn't the case before, either, and I don't want to complicate things even more).

t-vi committed:
clean up specialize_autogradzero

t-vi committed:
add test for optional list

t-vi committed:
Merge branch 'master' into optional_None

t-vi committed:
update hash for non-tensor args

t-vi committed:
remove CompleteArgumentSpec

t-vi committed:
fix C++ argspec test, infer CompleteTensorTpe in _jit_pass_complete_shape_analysis

t-vi commented:
So I think we end up with the same thing in the `_jit_pass_complete_shape_analysis`. There was a (pre-existing) bug in the ArgumentSpec test that it compared a `std::move`d  spec to a fresh one.
I could split out the removal of CompleteArgumentSpec if that's preferable. It's not strictly needed, but I felt bad having two differently working ArgumentSpec/CompleteArgumentSpec around when I needed to touch the first.

t-vi committed:
add TORCH_API for Windows

zdevito commented on test/cpp/jit/test_argument_spec.h:
```diff
@@ -37,6 +39,20 @@ void testArgumentSpec() {
   auto& GF = at::CUDA(at::kFloat);
   auto& GD = at::CUDA(at::kDouble);
 
+  auto graph = std::make_shared<Graph>();
```
Can you use `jit::compile(str)` to create the graph? We are trying to kill SymbolicVariable.

zdevito commented on torch/csrc/jit/argument_spec.cpp:
```diff
@@ -124,6 +132,10 @@ ArgumentSpec ArgumentSpecCreator::create(bool with_grad, const Stack& input)
         spec.addTensor(*stack[stack_top]++, with_grad);
         break;
+      case SPECIALIZE_NONTENSOR_OPTIONAL:
+        // consume a tensor and add to the argspec
```
comment is copy/paste error.

zdevito commented on torch/csrc/jit/argument_spec.h:
```diff
@@ -47,9 +53,15 @@ struct ArgumentInfo {
   }
 
  private:
+  unsigned is_nontensor_optional_ : 1;
```
The way this is encoded is pretty weird. It would be clearer as something like:

// Argspec has:

std::vector<ArgumentInfo> tensor_args; // tensor information recorded
std::vector<bool> optional_presence; // for each optional type true if present, false if not

```
// Related Argspec commands are:
SPECIALIZE_OPTIONAL_TENSOR, // push present/not-present to optional_presense, and if present, add the tensor information to tensor_args
SPECIALIZE_TENSOR, // push tensor information to tensor_args
SPECIALIZE_OPTIONAL, // push present/not-present to optional_presense, 
```

t-vi replied:
So elegant! Thanks!

zdevito commented on torch/csrc/jit/argument_spec.h:
```diff
@@ -163,183 +194,10 @@ struct ArgumentSpecCreator {
   std::vector<Inst> instructions_;
 };
 
-// CompleteArgumentSpec represents one particular specialization.
```
This code needs to be kept. It is being used in the TVM backend.

zdevito commented on torch/csrc/jit/operator.cpp:
```diff
@@ -457,10 +457,20 @@ bool Operator::matches(const Node* node) const {
     }
     TypePtr formal = *matched_type.type;
     if (!actuals[i]->type()->isSubtypeOf(formal)) {
+      if (actuals[i]->type() == NoneType::get() &&
```
same here

zdevito commented on torch/csrc/jit/passes/peephole.cpp:
```diff
@@ -188,16 +188,24 @@ void PeepholeOptimizeImpl(Block* block, bool addmm_fusion_enabled) {
+            node->owningGraph()->insertConstant(node->kind() == aten::__is__);
+        node->output()->replaceAllUsesWith(output);
+      } else {
+        for (size_t check_none_index : {0, 1}) {
```
I can't follow this code after a cursory reading. This needs simplification or comments.

t-vi replied:
It's unneeded now.

zdevito commented on torch/csrc/jit/passes/shape_analysis.cpp:
```diff
@@ -518,6 +518,19 @@ class ShapePropagator {
         }
         return;
       }
+      case prim::unchecked_unwrap_optional: {
```
Input to unchecked_unwrap_optional should _always_ be an optional type. Things will break if it isn't

t-vi replied:
Well, no. If we specialize optionals, it might be element type instead.

t-vi replied:
I changed it to test whether the input is a subtype of the output and if so, take the input type. This allows `Optional[Tensor]`s that have been specialized to `DimensionedTensorType` to have the more precise output type.

zdevito replied:
I think this is ok now.

zdevito commented on torch/csrc/jit/passes/shape_analysis.cpp:
```diff
@@ -540,10 +553,17 @@ class ShapePropagator {
       case aten::_unwrap_optional: {
-        auto input_ivalue = toIValue(node->input());
-        if (input_ivalue && input_ivalue->isNone()) {
+        // During analysis we don't want to pass None through here
```
same here

t-vi replied:
Again, it might be element (sub-) type after specialization.

zdevito commented on torch/csrc/jit/argument_spec.cpp:
```diff
@@ -186,6 +200,15 @@ std::vector<TypePtr> ArgumentSpecCreator::getSpecializedTypes(
+        auto ot = (*input_stack.back()++)->expect<OptionalType>();
+        auto& arg = spec.at(arg_spec_offset++);
+	if (arg.isNone()) {
+	  result_stack.back().emplace_back(NoneType::get());
```
This is the source of the problems with NoneType. Explicit None's are represented as:

```
a : Optional[int] = prim::Constant()
```
in the graph. NoneType is only used as an intermediate in the compiler. Instead of trying to change the input to NoneType. Instead I suggest disconnecting the input that must be none from its uses and replacing it with the appropriate None constant. This will require changing the API here:

ArgumentSpecCreator::specializeTypes(Graph& graph, ...) rather than simply returning new types. This change is fine to make, and will end up helping changes that I may need to do to argspec in the future.

t-vi replied:
This has been the thing I've been missing. Thank you so much!

The remaining not so nice thing here is that I don't know how to specialize Optionals to a Typed-None-Constant inside tuples and objects, because I cannot use `replaceAllUses`.
My tentative conclusion is that I cannot do this currently unless I can add some attribute to the optional type to indicate that it is actually a Typed-None constant.
What could be done is to see which uses are unpacking this, but that's really ugly and probably would remain incomplete.

zdevito reviewed (dismissed):
This is a really cool PR, but there are a couple of changes that introduce subtle bugs in the type system. See inline comments for how to avoid them.

t-vi commented:
Thanks @zdevito, it'll make a much shinier patch once I implement your advice.

So one thing I'd need a yes/no on is: Do you want `_jit_pass_complete_shape_analysis` to move to `ArgumentSpecCreator` (the "no" option being to continue to use `CompleteArgumentSpec`)?

t-vi committed:
Implement feedback from zdevito

t-vi committed:
clean unwrapping, tabs

t-vi commented:
So I think this is much better now. I left open the two things where I think I might not be completely following the advice to the letter (in the unwrapping) or see a limitation.

I left the `_jit_pass_complete_shape_analysis` with using the new mechanism and left `CompleteArgumentSpec` unchanged (except fixing one test).

t-vi committed:
Merge branch 'master' into optional_None

t-vi committed:
get_method? I don't know any methods. Oh, get_function you mean. You should say so.

zdevito commented:
@t-vi I've been swamped this week, so I haven't gotten to look at this change again, but I just wanted to let you know I haven't forgotten about and will try to get to it as soon as I can.

zdevito reviewed (commented):
This looks good! Thanks for the patience. I just have one minor question about the autogradzero optional, and one suggestion, otherwise this is ready to go.

t-vi commented on torch/csrc/jit/argument_spec.cpp:
```diff
@@ -169,13 +195,32 @@ std::vector<TypePtr> ArgumentSpecCreator::getSpecializedTypes(
+        auto& arg = spec.tensorAt(tensor_arg_spec_offset++);
+        if (!is_present) {
+          result_stack.back().emplace_back(input_type);
+        } else if (!arg.defined()) {
```
So for an earlier iteration @apaszke had the example of a function returning an `Optional[Tensor]`. On the other hand, if we now specialize `Optional[Tensor]`, we might not need it...

t-vi replied:
Seems OK without, so I replaced it with an assert.

t-vi committed:
Don't add None tensors to the argument stack

and assert that Optional tensors cannot be undefined.

Based on Zach's feedback. Thank you!

t-vi commented on torch/csrc/jit/argument_spec.cpp:
```diff
@@ -169,13 +195,32 @@ std::vector<TypePtr> ArgumentSpecCreator::getSpecializedTypes(
+      case SPECIALIZE_OPTIONAL_TENSOR: {
+        auto& input_type = *input_stack.back()++;
+        auto is_present = spec.isPresent(optional_arg_spec_offset++);
+        auto& arg = spec.tensorAt(tensor_arg_spec_offset++);
```
I fixed that, thanks!

t-vi commented:
If there is anything I can do to move this forward, I'd appreciate if you gave me a shout. Thank you!

eellison commented on torch/csrc/jit/argument_spec.cpp:
```diff
@@ -205,18 +207,17 @@ void ArgumentSpecCreator::specializeTypes(
         }
+        auto& arg = spec.tensorAt(tensor_arg_spec_offset++);
+        AT_ASSERT(arg.defined());
+        result_stack.back().emplace_back(DimensionedTensorType::create(
```
copy pasta from above

eellison reviewed (approved):
Looks like the comment has been addressed

wanchaol commented on torch/csrc/jit/init.cpp:
```diff
@@ -155,13 +155,22 @@ void initJITBindings(PyObject* module) {
-            AT_ASSERT(spec.size() == graph_inputs.size());
-            for (size_t i = 0; i < graph_inputs.size(); ++i) {
-              graph_inputs[i]->setType(spec.at(i));
+            ArgumentSpecCreator arg_spec_creator(*graph);
```
seems like this switching from completeArgSpec to only propagate dimension tensor information rather than complete tensors. I guess shape analysis works fine in this case, but does this means that we are missing the complete tensor type information? I would imagine if the following passes need the complete information then they could not find it.

t-vi replied:
Wouldn't the comment and code below this get back to CompleteTensorType and propagate those?

wanchaol replied:
oops you are right.. I was thinking `incompleteInferTypeFrom` does not get the complete information..

t-vi committed:
Merge branch 'master' into optional_None

t-vi committed:
Deduplicate dimensioned tensor type creation. Thanks @eelison for the suggestion

t-vi committed:
fix tests - apparently we have no graph after error

t-vi commented:
So I'm not super-happy with not getting a specialized graph for the failed script:
https://github.com/pytorch/pytorch/blob/dcbfa6759486567ee0e3de76c96c083340fad6d8/test/test_jit.py#L5566-L5571

wanchaol reviewed (approved):
lgtm

t-vi committed:
improve tests, remove fixme

t-vi commented:
So now I'm happy with the PyTorch test, but the caffe JIT CI tests are acting up. :(

t-vi committed:
Merge branch 'master' into optional_None

t-vi committed:
add WithInsertPoint

t-vi commented:
So finally it's all green again. :)

eellison commented:
oops didnt see u had imported it @wanchaol
